### PR TITLE
README rework

### PR DESCRIPTION
**Pull request by rricard**
The readme grew organically and forgot on the way to be a good intro to the proposal. This attempts to fix that.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,126 +61,131 @@ Deep equality as defined in user libraries can vary significantly, in part due t
...
-```
+#### `Tuple`
 
-Non-string keys are coerced to strings.
+Here is a tuple that can contain different values (of different types):
 
 ```js
-assert(#{ [true]: 1 } === #{ true: 1 })
-assert(#{ [true]: 1 } === #{ ["true"]: 1 })
+const document = [
```
Comment (nicolo-ribaudo) [inline README.md:125]:
```suggestion
const document = #[
```

File: README.md [hunk 2]
```diff
@@ -61,126 +61,131 @@ Deep equality as defined in user libraries can vary significantly, in part due t
...
 
-const [a, b] = #[1, 2];
-assert(a === 1);
-assert(b === 2);
-```
+function move(start, deltaX, deltaY) {
+  // we always return a tuple after moving
+  return #[
+    start.x + deltaX,
+    start.y + deltaY,
```
Comment (nicolo-ribaudo) [inline README.md:169]:
```suggestion
    start[0] + deltaX,
    start[1] + deltaY,
```

File: README.md [hunk 3]
```diff
@@ -61,126 +61,131 @@ Deep equality as defined in user libraries can vary significantly, in part due t
...
-const #{ a, b } = #{ a: 1, b: 2 };
-
-// SyntaxError
-const #[a, b] = #[1, 2];
-```
+See [more examples here](./details.md#more-exhaustive-tuple-manipulations).
 
 #### Forbidden cases
 
+As stated before Record & Tuple ar edeeply immutable: attempting to insert an object in them will result in a TypeError:
```
Comment (chicoxyzzy) [inline README.md:187]:
```suggestion
As stated before Record & Tuple are edeeply immutable: attempting to insert an object in them will result in a TypeError:
```

File: README.md [hunk 4]
```diff
@@ -61,126 +61,131 @@ Deep equality as defined in user libraries can vary significantly, in part due t
...
+console.log(document[1].keywords[1]); // tc39
 
-```js
-const formData = #{ title: "Implement all the things" }
-const taskNow = #{ id: 42, status: "WIP", ...formData }
-const taskLater = #{ ...taskNow, status: "DONE" }
-
-// A reminder: The ordering of keys in record literals does not affect equality (and is not retained)
-assert(taskLater === #{ status: "DONE", title: formData.title, id: 42 })
+// Finally you can also use the funtions available on the Tuple prototype
```
Comment (chicoxyzzy) [inline README.md:151]:
```suggestion
// Finally you can also use the functions available on the Tuple prototype
```

File: README.md [hunk 5]
```diff
@@ -658,6 +698,26 @@ This proposal is loosely related to a broader set of proposals, including [opera
 
 If we had user-defined value types, then it could make sense to use them in built-in features, such as [CSS Typed OM](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Typed_OM_API/Guide) or the [Temporal Proposal](https://github.com/tc39/proposal-temporal). However, this is far in the future, if it ever happens; for now, it works well to use objects for these sorts of features.
 
+## Are ECMAScript Record & Tuple the same as TypeScript's [Record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt) & [Tuple](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple)
+
+No they are completely different.
+
+Records in Typescript are a generic utility type to represent an object taking a key type matching with a value type. They still represents objects.
```
Comment (chicoxyzzy) [inline README.md:705]:
```suggestion
Records in Typescript are a generic utility type to represent an object taking a key type matching with a value type. They still represent objects.
```

File: README.md [hunk 6]
```diff
@@ -519,6 +524,14 @@ At a high level, the object/primitive distinction helps form a hard line between
 
 An alternative to implementing Record and Tuple as primitives would be to use [operator overloading](https://github.com/tc39/proposal-operator-overloading) to achieve a similar result, by implementing an overloaded abstract equality (`==`) operator that deeply compares objects. While this is possible, it doesn't satisfy the full use case, because operator overloading doesn't provide an override for the `===` operator. We want the strict equality (`===`) operator to be a reliable check of "identity" for objects and "observable value" (modulo -0/+0/NaN) for value types.
 
+Another option is to perform what is called _interning_: we track globally Record or Tuple objects and if we attempt to create a new one that happens to be identical to an existing Record object, we now reference this existing Record instead of creating a new one. This is essentially what the [polyfill](https://github.com/bloomberg/record-tuple-polyfill) does. We're now equating value and identity. This approach creates problems once we extend that behavior across multiple JavaScript contexts and wouldn't give deep immutability by nature.
+
+## Will developers be familiar with this new concept?
```
Comment (littledan) [inline README.md:529]: +1 to this section. Maybe we should also talk here about our plans to do research to validate this hypothesis?

File: README.md [hunk 7]
```diff
@@ -519,6 +524,14 @@ At a high level, the object/primitive distinction helps form a hard line between
 
 An alternative to implementing Record and Tuple as primitives would be to use [operator overloading](https://github.com/tc39/proposal-operator-overloading) to achieve a similar result, by implementing an overloaded abstract equality (`==`) operator that deeply compares objects. While this is possible, it doesn't satisfy the full use case, because operator overloading doesn't provide an override for the `===` operator. We want the strict equality (`===`) operator to be a reliable check of "identity" for objects and "observable value" (modulo -0/+0/NaN) for value types.
 
+Another option is to perform what is called _interning_: we track globally Record or Tuple objects and if we attempt to create a new one that happens to be identical to an existing Record object, we now reference this existing Record instead of creating a new one. This is essentially what the [polyfill](https://github.com/bloomberg/record-tuple-polyfill) does. We're now equating value and identity. This approach creates problems once we extend that behavior across multiple JavaScript contexts and wouldn't give deep immutability by nature.
```
[the remark this answers is not in the document]
Comment (rricard) [reply README.md]: I think it is worth mentioning and saying it is slow actually. That way we clear up why we just don't do what the polyfill is doing.

File: README.md [hunk 8]
```diff
@@ -636,7 +649,34 @@ We are developing the deep path properties proposal as a separate follow-on prop
...
+  // (Simplified; we may want to return an existing symbol if it's already there)
+  const sym = Symbol();
+  references.set(sym, obj);
+  return sym;
+}
+
+export function deref(sym) { return references.get(sym); }
+```
+
+</details>
```
[the remark this answers is not in the document]
Comment (rricard) [reply README.md:704]: The details don't show up by default, they are here for completeness. I think the example before is important to show what's possible

File: README.md [hunk 9]
```diff
@@ -532,6 +532,41 @@ Record & Tuple is built to interoperate with objects and arrays well: you can re
 
 Developers used to manipulating objects in an immutable manner (such as transforming pieces of Redux state) will be able to continue to do the same manipulations they used to do on objects and arrays, this time, with more guarantees.
 
+## Why are Record & Tuple not based on `.get()`/`.set()` methods like [Immutable.js](https://immutable-js.github.io/immutable-js/)?
+
+If we want to keep access to Record & Tuple similar to Objects and Arrays as described in the previous section, we can't rely on methods to perform that access. Doing so would require us to branch code when trying to create a "generic" function able to take Objects/Arrays/Records/Tuples. We also want to avoid an ecoystem split where libraries choose which structures they want to operate on.
```
Comment (ljharb) [inline README.md:528]: With the `items` proposal, this would no longer be true for Arrays/Tuples.

Comment (rricard) [reply README.md]: Agreed, that would make it possible for that operation to be generic. It does seem impossible to fix that with objects as adding accessor methods would possibly clash with keys.

**Comment by rricard** [comment]
merging those changes for now, will need to add a few other PRs to finish the rationale properly

File: README.md [hunk 10]
```diff
@@ -658,6 +723,26 @@ This proposal is loosely related to a broader set of proposals, including [opera
...
 
+## What's the relationship between this proposal's Record & Tuple and TypeScript's [Record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt) & [Tuple](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple)?
+
+Although both kinds of Records relate to Objects, and both kinds of Tuples relate to Arrays, that's about where the similarity ends.
+
+Records in Typescript are a generic utility type to represent an object taking a key type matching with a value type. They still represent objects.
+
+Likewise Tuples in Typescript are a notation to express types in an array of a limited size (starting with TypeScript 4.0 they have a [variadic form](https://github.com/microsoft/TypeScript/pull/39094)).
+
+TS Records or Tuples could end up be usable on ECMAScript Records and Tuples as well, for instance:
```
Comment (robpalme) [inline README.md]: typo: being

File: README.md [hunk 11]
```diff
@@ -658,6 +723,26 @@ This proposal is loosely related to a broader set of proposals, including [opera
...
+
+```ts
+// potential TS Record of an ECMAScript Record
+const record: RecordRecord<string, number> = #{
```
Comment (robpalme) [inline README.md]: Typo: RecordRecord

File: README.md [hunk 12]
```diff
@@ -658,6 +723,26 @@ This proposal is loosely related to a broader set of proposals, including [opera
...
+  foo: 1,
+  bar: 2,
+};
+
+// potential TS Tuple of an ECMAScript Tuple
+const tuple: #[number, string] = #[1, "foo"];
```
Comment (robpalme) [inline README.md]: I think we're trying to show Tuples with today's TypeScript, so we should drop the first type annotation # character. 

If TS had support for Tuples as types, the whole annotation would be redundant.

**Comment by rricard** [comment]
thanks for the feedback rob but I already had that refactored prior to merging, let me know if the new version needs refactoring

File: README.md (hunk 9 above)
Comment (littledan) [reply README.md]: Which proposal are you referring to? The `item` method?

Even if some APIs to access Tuples vs Arrays are generic, this proposal aims to have greater ergonomics by going for maximal compatibility. I hope this simplifies the mental model for programmers, to minimize the number of differences they have to remember.

**Review by muhdaizwan — APPROVED** [review]

**Pull request merged**